File: Phase_1/Leetcode/two_sum.py
"""
🧩 Problem: Two Sum

Given:
An array of integers nums and an integer target.

Task:
Return indices of the two numbers such that they add up to target.
You may assume that each input would have exactly one solution, and you may not use the same element twice
You can return the answer in any order.
🔹 Example 1
Input:
nums = [2,7,11,15], target = 9
Output:
[0,1]
Explanation:
Because nums[0] + nums[1] == 9, we return [0, 1].
🔹 Example 2
Input:
nums = [3,2,4], target = 6
Output:
[1,2]
🔹 Example 3
Input:
nums = [3,3], target = 6
Output:
[0,1]
⚙️ Constraints
2 <= nums.length <= 10^4
-10^9 <= nums[i] <= 10^9
-10^9 <= target <= 10^9
Only one valid answer exists.

"""
def two_sum(nums, target):
    seen = {}
    for index, num in enumerate(nums):
        diff = target - num
        if diff in seen:
            return [seen[diff], index]
        seen[num] = index

print(two_sum([1, 5, 4, 7], 6)) #returns 0, 1

File: DSA/array-problems/two_sum.cpp
#include <bits/stdc++.h>

using namespace std;
/*
 ✅ 1. Two Sum    *

 Problem:
 Given an array nums and a target value, return indices of the two numbers such that they add up to the target.

 Example:
 Input: nums = [2,7,11,15], target = 9
 Output: [0,1]
 Reason: 2 + 7 = 9
 */




// reusable version
vector<int> twoSum(vector<int>& nums, int target){
    unordered_map<int, int> map;

    for (int i = 0; i < nums.size(); i++){
        int diff = target - nums[i];
        if(map.find(diff) != map.end()){
            return {map[diff], i};
        }
        map[nums[i]] = i;
    }
    return {};
}



int main(){
    vector<int> nums = {2, 7, 11, 15};
    int target = 9;

    unordered_map<int, int> track;
    vector<int> result;

    for (int i = 0; i < nums.size(); i++ ){
        int diff = target - nums[i];
        if (track.find(diff) != track.end()){
            result = {track[diff], i};
            break;
        }
        track[nums[i]] = i;
    }

    cout << "Indics: " << result[0] << " " << result[1] << endl;

    vector<int> result_ = twoSum(nums, 9);
    cout << "Indices: " << result_[0] << " " << result_[1] << endl;




    return 0;
}
